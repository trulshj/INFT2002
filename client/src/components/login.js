// @flow

import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button, NavBar } from '../widgets';
import history from '../history';
import userService from '../user-service';

class Login extends Component {
  username = '';
  password = '';
  message = '';

  render() {
    // If we're already logged in jsut redirect to profile
    if (userService.user != '') {
      history.push('/profile');
    }
    return (
      <Card title="Login">
        <Column width={3}>
          <Form.Input
            id={'username'}
            value={this.username}
            type="text"
            placeholder="Username"
            onChange={(event) => {
              this.username = event.currentTarget.value;
            }}
          />
          <Form.Input
            id={'password'}
            value={this.password}
            type="password"
            placeholder="Password"
            onChange={(event) => {
              this.password = event.currentTarget.value;
            }}
          />
          <br />
          <Button.Success
            onClick={() => {
              userService.login(this.username, this.password);

              // Janky way of waiting for the server
              let i = 0;
              let interval = setInterval(() => {
                if (userService.user != '') {
                  clearInterval(interval);
                  history.push('/profile');
                }
                if (i > 5) {
                  clearInterval(interval);
                  this.message = '❌ Wrong password or username';
                }
                i++;
              }, 16);
            }}
          >
            Login
          </Button.Success>
          {this.message}
          <br />
          Don't have a user? Register <a href="#/register">here</a>!
        </Column>
      </Card>
    );
  }
}

export default Login;
