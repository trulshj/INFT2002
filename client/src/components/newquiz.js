// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button } from '../widgets';
import quizService, {
  type Category,
  type QuizQuestionOption,
  type QuizQuestion,
  type Quiz,
} from '../quiz-service';
import Dropdown from 'react-dropdown';

//Component for creating new quizzes
export class NewQuiz extends Component {
  quizName = '';
  categories: Category[] = [];

  render() {
    console.log(this.categories.category_name);
    return (
      <>
        <Card title="New Quiz">
          <Row>
            <Column width={2}>
              <Form.Label>Quiz Name:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.quizName}
                onChange={(event) => (this.quizName = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Category:</Form.Label>
            </Column>
            <Column>
              <select id="categoryValue">
                {this.categories.map((category) => (
                  <option
                    key={category}
                    value={this.category}
                    onChange={(event) =>
                      (this.category = event.currentTarget.value) + console.log('Hei')
                    }
                    placeholder="Select an option"
                  >
                    {category.category_name}
                  </option>
                ))}
              </select>
            </Column>
          </Row>
        </Card>
        <Button.Success
          onClick={() => {
            console.log(this.quizName);
            quizService

              .create(this.quizName, this.category)
              .then((quizId) => history.push('/newQuiz/' + quizId))
              .catch((error: Error) => Alert.danger('Error creating quiz: ' + error.message));
          }}
        >
          Create quiz
        </Button.Success>
      </>
    );
  }
  mounted() {
    quizService
      .getAllcategories()
      .then((categories) => (this.categories = categories))
      .catch((error: Error) => Alert.danger('Error getting quiz: ' + error.message));
  }
}

export default NewQuiz;

//Component for creating new questions for quizzes
export class NewQuizQuestions extends Component<{ match: { params: { id: number } } }> {
  quizquestion: QuizQuestion = { quizQuestionId: 0, quizId: 0, question: '' };
  quizquestionoption: QuizQuestionOption = {
    quizQuestionOptionId: 0,
    quizQuestionId: 0,
    questionAnswer: '',
    isCorrect: false,
  };

  render() {
    return (
      <>
        <Card title="Create Questions">
          <Row>
            <Column width={2}>
              <Form.Label>Question:</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.quizquestion.question}
                onChange={(event) => (this.quizquestion.question = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 1</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.quizquestionoption.questionAnswer}
                onChange={(event) =>
                  (this.quizquestionoption.questionAnswer = event.currentTarget.value)
                }
              />
              <Form.Checkbox
                checked={this.quizquestionoption.isCorrect}
                onChange={() => {}}
                disabled
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 2</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.quizquestionoption.questionAnswer}
                onChange={(event) =>
                  (this.quizquestionoption.questionAnswer = event.currentTarget.value)
                }
              />
              <Form.Checkbox
                checked={this.quizquestionoption.isCorrect}
                onChange={() => {}}
                disabled
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 3</Form.Label>
            </Column>
            <Column>
              <Form.Input
                type="text"
                value={this.quizquestionoption.questionAnswer}
                onChange={(event) =>
                  (this.quizquestionoption.questionAnswer = event.currentTarget.value)
                }
              />
              <Form.Checkbox
                checked={this.quizquestionoption.isCorrect}
                onChange={() => {}}
                disabled
              />
            </Column>
          </Row>
        </Card>
        <Button.Success onClick={() => history.push('/quizzes')}>Save</Button.Success>
      </>
    );
  }

  mounted() {}
}
