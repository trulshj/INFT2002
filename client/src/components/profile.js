// @flow

import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button, NavBar } from '../widgets';
import userService from '../user-service';

class Profile extends Component {
  username = '';
  password = '';

  render() {
    return (
      <Card title="Profile">
        <Row>
          <Column>Logged in as {userService.user}</Column>
        </Row>
      </Card>
    );
  }
}

export default Profile;
