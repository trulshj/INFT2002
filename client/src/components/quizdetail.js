import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button, NavBar, StarRating } from '../widgets';
import quizService, { type Quiz, type Category, type QuizQuestion } from '../quiz-service';
import { createHashHistory, Route } from 'history';
import { NavLink } from 'react-router-dom';
import ReactStars from 'react-rating-stars-component';



const history = createHashHistory(); // Use history.push(...) to programmatically change path


/**
 * Component for getting details of a spesific quiz
 * 
 */
export class QuizDetail extends Component<{ match: { params: { quizId: number } } }> {
    quiz: Quiz = { quizId: 0, quizName: '', quizCategory: '' };
    questions: QuizQuestion[] = [];
  
    render() {
      console.log(this.quiz);
      return (
        <>
          <Card title={this.quiz.quiz_name}>
            {this.questions.map((question) => (
              <Card>
                <Row key={question.quizQuestionId}>
                  <Column width={10}>
                    <NavLink to={'/quizzes/' + question.quiz_id + '/' + question.quiz_question_id}>
                      {question.question}
                    </NavLink>
                  </Column>
                </Row>
              </Card>
            ))}
          </Card>
          <Card title="Rate quiz">
          <StarRating onChange={() => {
            quizService
              .createRating(this.props.rating, this.props.match.params.quizId)
              .then(alert(this.props.rating, this.props.match.params.quizId))
              .catch((error: Error) => Alert.danger('Error creating rating: ' + error.message));
          }}
        ></StarRating>
          </Card>
          
        </>
      );
    }
  
    mounted() {
      quizService
        .get(this.props.match.params.quizId)
        .then((quiz) => (this.quiz = quiz))
        .catch((error: Error) => Alert.danger('Error getting quiz: ' + error.message));
      quizService
        .getAllQuestionsInQuiz(this.props.match.params.quizId)
        .then((quizQuestions) => (this.questions = quizQuestions))
        .catch((error: Error) => Alert.danger('Error getting quiz questions: ' + error.message));
    }
  }

/**
 * Component for getting details of questions
 */
export class QuestionDetail extends Component<{ match: { params: { quizQuestionId: number } } }> {
    question: QuizQuestion = { quizQuestionId: 0, quizId: 0, question: '' };
    questionOption: QuizQuestionOption[] = [];
    questionOptionCorrect: QuizQuestionOption[] = [];
  
    render() {
      console.log(this.question);
      return (
        <>
          <Card title={this.question.question}>
            {this.questionOption.map((option) => (
              <Card>
                <Row key={option.quizQuestionOptionId}>
                  <Column width={10}>{option.question_answer}</Column>
                </Row>
              </Card>
            ))}
            <Card title="Correct Answer(s)">
              {this.questionOptionCorrect.map((option) => (
                <Card>
                  <Row key={option.quizQuestionOptionId}>
                    <Column width={10}>{option.question_answer}</Column>
                  </Row>
                </Card>
              ))}
            </Card>
          </Card>
        </>
      );
    }
  
    mounted() {
      quizService
        .getQuestion(this.props.match.params.quizQuestionId)
        .then((question) => (this.question = question))
        .catch((error: Error) => Alert.danger('Error getting question: ' + error.message));
      quizService
        .getQuestionOption(this.props.match.params.quizQuestionId)
        .then((questionOption) => (this.questionOption = questionOption))
        .catch((error: Error) => Alert.danger('Error getting question options: ' + error.message));
      quizService
        .getQuestionOptionCorrect(this.props.match.params.quizQuestionId)
        .then((questionOptionCorrect) => (this.questionOptionCorrect = questionOptionCorrect))
        .catch((error: Error) =>
          Alert.danger('Error getting correct question option: ' + error.message),
        );
    }
  }
  