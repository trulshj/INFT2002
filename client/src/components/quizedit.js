import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button, NavBar } from '../widgets';
import quizService, { type Quiz, type Category, type QuizQuestion } from '../quiz-service';
import { createHashHistory, Route } from 'history';
import { NavLink } from 'react-router-dom';

const history = createHashHistory(); // Use history.push(...) to programmatically change path

/**
 * Component for editing quiz
 * TODO:
 */

export class QuizEdit extends Component<{ match: { params: { quizId: number } } }> {
  quiz: Quiz = { quizId: 0, quizName: '', quizCategory: '' };
  questions: QuizQuestion[] = [];
  categories: Category[] = [];

  render() {
    //console.log(this.quiz);
    return (
      <>
        <Card>
          <Card>
            <Row>
              <Column width={1}>Edit Quiz title:</Column>
              <Column width={4}>
                <Form.Input
                  id="quizTitle"
                  type="text"
                  value={this.quiz.quiz_name}
                  onChange={(event) => (this.quiz.quiz_name = event.currentTarget.value)}
                />
              </Column>
              <Column>
                <Form.Select
                  id="categoryValue"
                  onChange={(event) => (this.quiz.quiz_category = event.currentTarget.value)}
                  value={this.quiz.quiz_category}
                >
                  {this.categories.map((category) => (
                    <option
                      key={category.category_name}
                      value={category.category_name}
                      placeholder="Select an option"
                    >
                      {category.category_name}
                    </option>
                  ))}
                </Form.Select>
              </Column>
            </Row>
          </Card>
          {this.questions.map((question) => (
            <Card key={question.quiz_question_id}>
              <Row>
                <Column width={10}>
                  <NavLink
                    to={'/quizzes/' + question.quiz_id + '/' + question.quiz_question_id + '/edit'}
                  >
                    {question.question}
                  </NavLink>
                </Column>
              </Row>
            </Card>
          ))}
        </Card>
        <Card>
          <Button.Success
            onClick={() => history.push('/quizzes/' + this.quiz.quiz_id + '/edit/addQuestion')}
          >
            Add question
          </Button.Success>
          <Button.Light
            onClick={() =>
              quizService
                .updateQuiz(this.quiz)
                .then(() => history.push('/quizzes'))
                .catch((error: Error) => Alert.danger('Error updating quiz: ' + error.message))
            }
          >
            Save changes
          </Button.Light>
          <Button.Danger
            onClick={() => {
              this.questions.map((question) =>
                quizService
                  .deleteOption(question.quiz_question_id)
                  .catch((error: Error) =>
                    Alert.danger('Error deleting options for questions for quiz'),
                  ),
              );
              quizService
                .deleteQuizQuestions(this.quiz.quiz_id)
                .catch((error: Error) =>
                  Alert.danger('Error deleting questions for quiz: ' + error.message),
                );
              quizService
                .deleteRating(this.quiz.quiz_id)
                .catch((error: Error) => Alert.danger('Error deleting ratings for quiz'));
              quizService
                .deleteQuiz(this.quiz.quiz_id)
                .then(() => history.push('/quizzes'))
                .catch((error: Error) => Alert.danger('Error deleting quiz: ' + error.message));
            }}
          >
            Delete quiz
          </Button.Danger>
        </Card>
      </>
    );
  }

  mounted() {
    quizService
      .get(this.props.match.params.quizId)
      .then((quiz) => (this.quiz = quiz))
      .catch((error: Error) => Alert.danger('Error getting quiz: ' + error.message));
    quizService
      .getAllQuestionsInQuiz(this.props.match.params.quizId)
      .then((quizQuestions) => (this.questions = quizQuestions))
      .catch((error: Error) => Alert.danger('Error getting quiz questions: ' + error.message));
    quizService
      .getAllcategories()
      .then((categories) => {
        this.categories = categories;
      })
      .catch((error: Error) => Alert.danger('Error getting categories: ' + error.message));
  }
}

/**
 * Component for editing questions
 * TODO:
 *  */

export class QuestionEdit extends Component<{ match: { params: { quizQuestionId: number } } }> {
  question: QuizQuestion = { quizQuestionId: 0, quizId: 0, question: '' };
  questionOption: QuizQuestionOption[] = [];

  render() {
    //(this.question);
    //console.log(this.questionOption);
    return (
      <>
        <Card>
          <Card>
            <Row>
              <Column width={1}>Edit Question title:</Column>
              <Column width={4}>
                <Form.Input
                  id="questionTitle"
                  type="text"
                  value={this.question.question}
                  onChange={(event) => (this.question.question = event.currentTarget.value)}
                />
              </Column>
            </Row>
          </Card>
          {this.questionOption.map((option) => (
            <Card key={option.quiz_question_option_id}>
              <Row>
                <Column width={4}>
                  <Form.Input
                    id="optionAnswer"
                    type="text"
                    value={option.question_answer}
                    onChange={(event) => (option.question_answer = event.currentTarget.value)}
                  />
                </Column>
                <Column>
                  <Form.Checkbox
                    id="optionCheckbox"
                    checked={option.is_correct}
                    onChange={(event) => (option.is_correct = event.currentTarget.checked)}
                  />{' '}
                  <Form.Label>Correct answer</Form.Label>
                </Column>
                <Column>
                  <Button.Success
                    id="optionUpdateButton"
                    onClick={() =>
                      quizService
                        .updateOption(option)
                        //.then(() => console.log(option))
                        .catch((error: Error) => Alert.danger('Error editing option: ' + error))
                    }
                  >
                    Save
                  </Button.Success>
                </Column>
              </Row>
            </Card>
          ))}
        </Card>
        <Card>
          <Button.Light
            onClick={() =>
              quizService
                .updateQuestion(this.question)
                .then(() => history.push('/quizzes/' + this.question.quiz_id + '/edit'))
                .catch((error: Error) => Alert.danger('Error updating question: ' + error.message))
            }
          >
            Save changes
          </Button.Light>
          <Button.Danger
            onClick={() => {
              quizService
                .deleteOption(this.question.quiz_question_id)
                .catch((error: Error) =>
                  Alert.danger('Error deleting options for questions for quiz'),
                );
              quizService
                .deleteQuestion(this.question.quiz_question_id)
                .then(() => history.push('/quizzes/' + this.question.quiz_id + '/edit'))
                .catch((error: Error) => Alert.danger('Error deleting question: ' + error.message));
            }}
          >
            Delete question
          </Button.Danger>
        </Card>
      </>
    );
  }

  mounted() {
    quizService
      .getQuestion(this.props.match.params.quizQuestionId)
      .then((question) => (this.question = question))
      .catch((error: Error) => Alert.danger('Error getting question: ' + error.message));
    quizService
      .getQuestionOption(this.props.match.params.quizQuestionId)
      .then((questionOption) => (this.questionOption = questionOption))
      .catch((error: Error) => Alert.danger('Error getting question options: ' + error.message));
    quizService
      .getQuestionOptionCorrect(this.props.match.params.quizQuestionId)
      .then((questionOptionCorrect) => (this.questionOptionCorrect = questionOptionCorrect))
      .catch((error: Error) =>
        Alert.danger('Error getting correct question option: ' + error.message),
      );
  }
}

/**
 * Component for adding new questions to existing quizzes
 */
export class QuizAddQuestion extends Component<{ match: { params: { quizId: number } } }> {
  quiz: Quiz = { quizId: 0, quizName: 'test', quizCategory: '' };
  question = '';
  option1 = '';
  isCorrect1 = false;
  option2 = '';
  isCorrect2 = false;
  option3 = '';
  isCorrect3 = false;

  render() {
    return (
      <>
        <Card title="Create Questions for quiz: ">
          {this.quiz.quiz_name}
          <Row>
            <Column width={2}>
              <Form.Label>Question</Form.Label>
            </Column>
            <Column width={4}>
              <Form.Input
                id="question_id"
                type="text"
                value={this.question}
                onChange={(event) => (this.question = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 1</Form.Label>
            </Column>
            <Column width={4}>
              <Form.Input
                id="option1"
                type="text"
                value={this.option1}
                onChange={(event) => (this.option1 = event.currentTarget.value)}
              />
            </Column>
            <Column>
              <Form.Checkbox
                id="isCorrect1"
                checked={this.isCorrect1}
                onChange={(event) => (this.isCorrect1 = event.currentTarget.checked)}
              />
              <Form.Label>Correct answer</Form.Label>
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 2</Form.Label>
            </Column>
            <Column width={4}>
              <Form.Input
                id="option2"
                type="text"
                value={this.option2}
                onChange={(event) => (this.option2 = event.currentTarget.value)}
              />
            </Column>
            <Column>
              <Form.Checkbox
                id="isCorrect2"
                checked={this.isCorrect2}
                onChange={(event) => (this.isCorrect2 = event.currentTarget.checked)}
              />
              <Form.Label>Correct answer</Form.Label>
            </Column>
          </Row>
          <Row>
            <Column width={2}>
              <Form.Label>Option 3</Form.Label>
            </Column>
            <Column width={4}>
              <Form.Input
                id="option3"
                type="text"
                value={this.option3}
                onChange={(event) => (this.option3 = event.currentTarget.value)}
              />
            </Column>
            <Column>
              <Form.Checkbox
                id="isCorrect3"
                checked={this.isCorrect3}
                onChange={(event) => (this.isCorrect3 = event.currentTarget.checked)}
              />
              <Form.Label>Correct answer</Form.Label>
            </Column>
          </Row>
        </Card>
        <Button.Success
          onClick={() => {
            quizService
              .createQuestion(
                this.quiz.quiz_id,
                this.question,
                this.option1,
                this.isCorrect1,
                this.option2,
                this.isCorrect2,
                this.option3,
                this.isCorrect3,
              )
              .then(
                alert('Tnx for creating Question. ' + this.question + '. Go ahead and add more!'),
              )
              .then(history.push('/quizzes/' + this.quiz.quiz_id + '/edit'))
              .catch((error: Error) => Alert.danger('Error creating question: ' + error.message));
            (this.question = ''),
              (this.option1 = ''),
              (this.isCorrect1 = false),
              (this.option2 = ''),
              (this.isCorrect2 = false),
              (this.option3 = ''),
              (this.isCorrect3 = false);
          }}
        >
          Add question
        </Button.Success>
      </>
    );
  }

  mounted() {
    quizService
      .get(this.props.match.params.quizId)
      .then((quiz) => (this.quiz = quiz))
      .catch((error: Error) => Alert.danger('Error getting quiz: ' + error.message));
  }
}
