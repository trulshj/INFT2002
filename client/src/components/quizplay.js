// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Form, Button, NavBar } from '../widgets';
import quizService, { type Quiz, type Category, type QuizQuestion } from '../quiz-service';
import playService from '../play-service';
import { createHashHistory } from 'history';
import { NavLink } from 'react-router-dom';
import ReactStars from 'react-rating-stars-component';

const history = createHashHistory();

export class QuizPlay extends Component<{ match: { params: { quizId: number } } }> {
  quiz: Quiz = { quizId: 0, quizName: '', quizCategory: '', rating: 0 };
  // question: QuizQuestion = { quizQuestionId: 0, quizId: 0, question: '' };
  questionOption: QuizQuestionOption[] = [];
  questions: QuizQuestion[] = [];
  // Is the quiz finished?
  quizFinished = false;
  // Points you get in the quiz
  points: number = 0;
  // Star rating
  rating = 0;

  render() {
    return (
      <>
        <Row>
          <Column width={5}>
            <h3>Questions:</h3>
          </Column>
          <Column width={5}>{this.quizFinished ? <h3>Answers:</h3> : null}</Column>
        </Row>
        {this.questions.map((question) => (
          <Row key={question.quiz_question_id}>
            <Column width={5}>
              <PlayOptions quizQuestionId={question.quiz_question_id} />
            </Column>
            <Column width={5}>
              {this.quizFinished ? (
                <PlayOptionAnswer quizQuestionId={question.quiz_question_id} />
              ) : null}
            </Column>
          </Row>
        ))}
        {!this.quizFinished ? (
          <Button.Success
            id="buttonFinishQuiz"
            onClick={() => {
              this.points = 0;
              if (!this.quizFinished) {
                this.quizFinished = true;
              }

              this.points = playService.checkAnswers();
            }}
          >
            Submit Answer
          </Button.Success>
        ) : (
          <Card>
            <h2>{`You got ${this.points} points!`}</h2>
          </Card>
        )}
        {this.quizFinished ? (
          <Card title="Rate quiz">
            <ReactStars
              id="ratingStars"
              size={24}
              value={this.rating}
              onChange={(newRating) => {
                this.rating = newRating;
                quizService
                  .createRating(this.rating, this.props.match.params.quizId)
                  .catch((error: Error) => Alert.danger('Error creating rating: ' + error.message));
              }}
            ></ReactStars>
          </Card>
        ) : null}
      </>
    );
  }

  mounted() {
    quizService
      .get(this.props.match.params.quizId)
      .then((quiz) => (this.quiz = quiz))
      .catch((error: Error) => Alert.danger('Error getting quiz: ' + error.message));
    quizService
      .getAllQuestionsInQuiz(this.props.match.params.quizId)
      .then((quizQuestions) => (this.questions = quizQuestions))
      .catch((error: Error) => Alert.danger('Error getting quiz questions: ' + error.message));
  }
}

export class PlayOptions extends Component {
  question: QuizQuestion = { quiz_question_id: 0, quizId: 0, question: '' };
  questionOption: QuizQuestionOption[] = [];
  questionOptionCorrect: QuizQuestionOption[] = [];

  render() {
    return (
      <>
        <Card title={this.question.question}>
          {this.questionOption.map((option) => (
            <Card key={option.quiz_question_option_id}>
              <Row>
                <Column width={10}>{option.question_answer}</Column>
                <Form.Checkbox
                  id={option.quiz_question_option_id}
                  onChange={(event) => {
                    event.currentTarget.checked = event.currentTarget.checked;

                    playService.setAnswer(
                      this.props.quizQuestionId,
                      (event.currentTarget.id + 2) % 3,
                      event.currentTarget.checked,
                    );
                  }}
                />
              </Row>
            </Card>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    quizService
      .getQuestion(this.props.quizQuestionId)
      .then((question) => (this.question = question))
      .catch((error: Error) => Alert.danger('Error getting question: ' + error.message));
    quizService
      .getQuestionOption(this.props.quizQuestionId)
      .then((questionOption) => (this.questionOption = questionOption))
      .catch((error: Error) => Alert.danger('Error getting question options: ' + error.message));
    quizService
      .getQuestionOptionCorrect(this.props.quizQuestionId)
      .then((questionOptionCorrect) => (this.questionOptionCorrect = questionOptionCorrect))
      .catch((error: Error) =>
        Alert.danger('Error getting correct question option: ' + error.message),
      );

    setTimeout(() => {
      const id = this.question.quiz_question_id;
      let corr = [];
      for (const q of this.questionOption) {
        if (q.quiz_question_id == id) {
          corr.push(q.is_correct);
        }
      }
      const ans = { questionId: id, givenAnswers: [false, false, false], correctAnswers: corr };
      playService.addQuestionAnswers(ans);
    }, 100);
  }
}

export class PlayOptionAnswer extends Component {
  question: QuizQuestion = { quizQuestionId: 0, quizId: 0, question: '' };
  questionOption: QuizQuestionOption[] = [];
  questionOptionCorrect: QuizQuestionOption[] = [];

  render() {
    return (
      <>
        <Card title={this.question.question}>
          {this.questionOption.map((option) => (
            <Card key={option.quiz_question_option_id}>
              <Row>
                <Column width={10}>{option.question_answer}</Column>
                <Form.Checkbox checked={option.is_correct} readOnly={true} />
              </Row>
            </Card>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    quizService
      .getQuestion(this.props.quizQuestionId)
      .then((question) => (this.question = question))
      .catch((error: Error) => Alert.danger('Error getting question: ' + error.message));
    quizService
      .getQuestionOption(this.props.quizQuestionId)
      .then((questionOption) => (this.questionOption = questionOption))
      .catch((error: Error) => Alert.danger('Error getting question options: ' + error.message));
    quizService
      .getQuestionOptionCorrect(this.props.quizQuestionId)
      .then((questionOptionCorrect) => (this.questionOptionCorrect = questionOptionCorrect))
      .catch((error: Error) =>
        Alert.danger('Error getting correct question option: ' + error.message),
      );
  }
}
