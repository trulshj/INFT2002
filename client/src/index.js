// @flow

import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, Row, Column, Form, Button } from './widgets';
import Login from './components/login';
import NewQuiz from './components/newquiz';
import Quizzes from './components/quizzes';
import Register from './components/register';

class Menu extends Component {
  render() {
    return (
      <NavBar brand="Quiz App">
        <NavBar.Link to="/quizzes">Quizzes</NavBar.Link>
        <NavBar.Link to="/newQuiz">New Quiz</NavBar.Link>
        <NavBar.Link to="/login">Login</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  render() {
    return <Card title="Welcome">Quiz App for Quizzes</Card>;
  }
}

const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        <Route exact path="/quizzes" component={Quizzes} />
        <Route exact path="/newQuiz" component={NewQuiz} />
        <Route exact path="/login" component={Login} />
        <Route exact path="/register" component={Register} />
      </div>
    </HashRouter>,
    root,
  );
