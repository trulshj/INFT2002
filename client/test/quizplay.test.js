// @flow

import * as React from 'react';
import { QuizPlay, PlayOptions, PlayOptionAnswer } from '../src/components/quizplay';
import quizService, { type Quiz, type Category, type QuizQuestion } from '../src/quiz-service';
import { shallow } from 'enzyme';
import { Card, Alert, Row, Column, Form, Button, NavBar, StarRating } from '../src/widgets';
import { NavLink, HashRouter, Route } from 'react-router-dom';
import ReactStars from 'react-rating-stars-component';


jest.mock('../src/quiz-service', () => {
    class QuizService {
      get(quizId: number) {
        return Promise.resolve({ quiz_id: 1, quiz_name: 'Land i Europa', quiz_category: 'Geografi' });
      }
  
      getAllQuestionsInQuiz(quizId: number) {
        return Promise.resolve([
          { quiz_question_id: 1, quiz_id: 1, question: 'Hva heter hovedstaden i Norge?' },
          { quiz_question_id: 2, quiz_id: 1, question: 'Hva heter hovedstaden i Sverige?' },
        ]);
      }
  
      getAllcategories() {
        return Promise.resolve([{ category_name: 'Geografi' }, { category_name: 'Kultur' }]);
      }
  
      getAll() {
        return Promise.resolve([
          { quiz_id: 1, quiz_name: 'Land i Europa', quiz_category: 'Geografi' },
          { quiz_id: 2, quiz_name: 'Tradisjoner i Norge', quiz_category: 'Kultur' },
        ]);
      }
  
      getQuestion(quizQuestionId: number) {
        return Promise.resolve({
          quiz_question_id: 1,
          quiz_id: 1,
          question: 'Hva heter hovedstaden i Norge?',
        });
      }
  
      getQuestionOption(quizQuestionId: number) {
        return Promise.resolve([
          { quiz_question_option_id: 1, 
            quiz_question_id: 1, 
            question_answer: 'Oslo', 
            is_correct: 1 },
          {
            quiz_question_option_id: 2,
            quiz_question_id: 1,
            question_answer: 'Sverige',
            is_correct: 0,
          },
          {
            quiz_question_option_id: 3,
            quiz_question_id: 1,
            question_answer: 'Moskva',
            is_correct: 0,
          },
        ]);
      }
      getQuestionOptionCorrect(quizQuestionId: number) {
        return Promise.resolve({
          quiz_question_option_id: 1,
          quiz_question_id: 1,
          question_answer: 'Oslo',
          is_correct: 1,
        });
      }
    }
    return new QuizService();
  });


//Test for NavBar
describe('QuizPlay', () => {

    describe('QuizPlay tests', () => {
      test('PlayOptions draws correctly', (done) => {
        const wrapper = shallow(<QuizPlay match={{ params: { quizQuestionId: 1 } }} />);
    
        setTimeout(() => {
          expect(
            wrapper.containsAllMatchingElements([
            <Column width={5}></Column>,
            ]),
          ).toEqual(true);
          done();
        });
      });

      test('PlayOptions draws correctly', (done) => {
        const wrapper = shallow(<PlayOptions match={{ params: { quizQuestionId: 1 } }} />);
    
        setTimeout(() => {
          expect(
            wrapper.containsAllMatchingElements([
            <Card title="Hva heter hovedstaden i Norge?"></Card>,
              <Column value="Oslo"></Column>,
              <Column value="Sverige"></Column>,
              <Column value="Moskva"></Column>,
            ]),
          ).toEqual(true);
          done();
        });
      });

      test('PlayOptionAnswer draws correctly', (done) => {
        const wrapper = shallow(<PlayOptionAnswer match={{ params: { quizQuestionId: 1 } }} />);
    
        setTimeout(() => {
          expect(
            wrapper.containsAllMatchingElements([
              <Card title="Hva heter hovedstaden i Norge?"></Card>,
              <Column value="Oslo"></Column>,
              <Column value="Sverige"></Column>,
              <Column value="Moskva"></Column>,
            ]),
          ).toEqual(true);
          done();
        });
      });
  });
});