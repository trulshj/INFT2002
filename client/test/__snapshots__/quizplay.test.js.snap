// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayOptionAnswer PlayOptionAnswer draws consitently 1`] = `
<Fragment>
  <Card
    title="Hva heter hovedstaden i Norge?"
  >
    <Card
      key="1"
    >
      <Row>
        <Column
          width={10}
        >
          Oslo
        </Column>
        <FormCheckbox
          checked={1}
          readOnly={true}
        />
      </Row>
    </Card>
    <Card
      key="2"
    >
      <Row>
        <Column
          width={10}
        >
          Sverige
        </Column>
        <FormCheckbox
          checked={0}
          readOnly={true}
        />
      </Row>
    </Card>
    <Card
      key="3"
    >
      <Row>
        <Column
          width={10}
        >
          Moskva
        </Column>
        <FormCheckbox
          checked={0}
          readOnly={true}
        />
      </Row>
    </Card>
  </Card>
</Fragment>
`;

exports[`PlayOptions tests PlayOptions draws consitently 1`] = `
<Fragment>
  <Card
    title="Hva heter hovedstaden i Norge?"
  >
    <Card
      key="1"
    >
      <Row>
        <Column
          width={10}
        >
          Oslo
        </Column>
        <FormCheckbox
          id={1}
          onChange={[Function]}
        />
      </Row>
    </Card>
    <Card
      key="2"
    >
      <Row>
        <Column
          width={10}
        >
          Sverige
        </Column>
        <FormCheckbox
          id={2}
          onChange={[Function]}
        />
      </Row>
    </Card>
    <Card
      key="3"
    >
      <Row>
        <Column
          width={10}
        >
          Moskva
        </Column>
        <FormCheckbox
          id={3}
          onChange={[Function]}
        />
      </Row>
    </Card>
  </Card>
</Fragment>
`;

exports[`QuizPlay tests QuizPlay draws consitently 1`] = `
<Fragment>
  <Row>
    <Column
      width={5}
    >
      <h3>
        Questions:
      </h3>
    </Column>
    <Column
      width={5}
    />
  </Row>
  <Row
    key="1"
  >
    <Column
      width={5}
    >
      <PlayOptions
        quizQuestionId={1}
      />
    </Column>
    <Column
      width={5}
    />
  </Row>
  <Row
    key="2"
  >
    <Column
      width={5}
    >
      <PlayOptions
        quizQuestionId={2}
      />
    </Column>
    <Column
      width={5}
    />
  </Row>
  <ButtonSuccess
    id="buttonFinishQuiz"
    onClick={[Function]}
  >
    Submit Answer
  </ButtonSuccess>
</Fragment>
`;
